### PR TITLE
Trunk-split decoder: specialized surface/volume transformer branches

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Trunk-split surface decoder: split the transformer TRUNK (not the output head) to create specialized surface vs. volume processing pathways, targeting the wsy/wsz binding gap.**

**Context:** nezuko #347 tested splitting the *output head* into separate surface/volume decoders. This backfired — it added parameters but hurt metrics by removing shared surface/volume representation. The lesson: head-split after a shared trunk doesn't work.

**New approach:** Split the trunk itself. Use a shared lower trunk (layers 1–3) and specialized upper trunk branches (layers 4–5) for surface vs. volume. The surface branch sees only surface-relevant tokens at training; the volume branch sees only volume tokens. Both branches share the same lower-level representations (shape, global geometry) but can specialize their higher-level features.

**Why this should work:** Wall shear stress (wsy/wsz) is a surface-only quantity requiring boundary-layer physics understanding. Volume pressure is interior. A single trunk must learn both representations simultaneously, creating representational competition. Splitting the upper trunk allows specialized gradient flow for each domain while preserving shared low-level features.

**Architecture:**
```
Input → FourierPE → layers 1-3 (shared trunk)
                    ↓                    ↓
          layers 4-5 (surface branch)  layers 4-5 (volume branch)
                    ↓                    ↓
          surface head                volume head
          (cp, wsx, wsy, wsz)        (vol_p)
```

## Instructions

**Add `--trunk-split-layer` flag to `train.py`** (default: None/disabled). When set to an integer `N` (e.g., 3), layers 1..N are shared and layers N+1..end are duplicated into surface/volume branches.

### 1. Implement trunk-split architecture

In `train.py` (model definition section), modify the transformer to support branching:

```python
class SplitTrunkModel(nn.Module):
    def __init__(self, ..., trunk_split_layer=None):
        ...
        # Shared lower trunk
        self.shared_blocks = nn.ModuleList([TransformerBlock(...) for _ in range(trunk_split_layer)])
        # Surface-specialized upper trunk
        self.surface_blocks = nn.ModuleList([TransformerBlock(...) for _ in range(n_layers - trunk_split_layer)])
        # Volume-specialized upper trunk
        self.volume_blocks = nn.ModuleList([TransformerBlock(...) for _ in range(n_layers - trunk_split_layer)])
        # Separate heads for surface and volume
        self.surface_head = ...  # predicts cp, wsx, wsy, wsz
        self.volume_head = ...   # predicts vol_p

    def forward(self, x, ...):
        # Shared lower trunk
        for block in self.shared_blocks:
            x = block(x)
        # Surface branch
        x_surf = x.clone()
        for block in self.surface_blocks:
            x_surf = block(x_surf)
        # Volume branch
        x_vol = x.clone()
        for block in self.volume_blocks:
            x_vol = block(x_vol)
        # Decode
        surface_out = self.surface_head(x_surf)  # cp, wsx, wsy, wsz
        volume_out = self.volume_head(x_vol)      # vol_p
        return combine(surface_out, volume_out)
```

**Split at layer 3 out of 5** (i.e., `--trunk-split-layer 3`). This keeps 60% shared, 40% specialized.

### 2. Training run

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --no-compile-model \
  --trunk-split-layer 3 \
  --kill-thresholds "500:train/loss<5,178000:val_primary/abupt_axis_mean_rel_l2_pct<10" \
  --wandb-group bengio-wave20-trunk-split \
  --wandb-name violet/trunk-split-layer3 \
  --agent violet
```

If ep10 val_abupt > 8.5%, abort. If ep10 val_abupt < 8.5%, run to ep50. Report per-epoch val metrics for abupt, surface_pressure, wsy, wsz.

**Important note on head-split:** nezuko #347 tested head-only split (after shared trunk) → failed. This experiment tests trunk-split (split the trunk itself). These are fundamentally different — ensure your implementation splits transformer layers, not just output projections.

## Baseline

Current best (PR #174, alphonse, val metrics at ep~45.3, step 807,025, run `vu4jsiic`):

| Metric | Current Best (val) | AB-UPT Target | Gap |
|--------|-------------------|--------------|-----|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **6.9549** | 4.51 | 2.44pp |
| `val_primary/surface_pressure_rel_l2_pct` | 4.5644 | 3.82 | 0.74pp |
| `val_primary/volume_pressure_rel_l2_pct` | **3.9361** ✓ | 6.08 | solved |
| `val_primary/wall_shear_y_rel_l2_pct` | 8.7345 | 3.65 | **5.08pp** |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.5766 | 3.63 | **6.95pp** |

Reproduce baseline:
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --fourier-pe --no-use-ema --lr 3e-4 --lr-cosine-t-max 50 \
  --no-compile-model --wandb-group bengio-wave9
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Acknowledged — hypothesis received. Trunk-split decoder with specialized surface/volume transformer branches is an architectural hypothesis targeting the fact that surface (wsy/wsz) and volume (vol_p) predictions have very different error profiles. Current baseline to beat: **val_abupt = 6.9549%** (PR #174, run `vu4jsiic`, ep~45.3).

**Kill gates (abupt):** EP5 > 13%, EP10 > 10.5%, EP15 > 8.5%, EP20 > 8.0%.

Architecture change risk: make sure parameter count is comparable to baseline (~3.99M params) — a larger model that beats baseline only due to capacity is less informative. Report n_params alongside results.

Post your W&B run ID(s) and first val checkpoint when training starts.

**Comment by morganmcg1** [comment]
STUDENT violet: This is a follow-up on your assigned PR. Please confirm you have started your training run and post your W&B run ID(s). The baseline to beat is **val_abupt = 6.9549%** (PR #174, run `vu4jsiic`). Post ep1 metrics as soon as available and apply standard kill gates (ep5: >13%, ep10: >10.5%, ep15: >8.5%, ep20: >8.0%).

**Comment by morganmcg1** [comment]
ADVISOR (infra-block escalation, 2026-05-03): Pod `senpai-bengio-violet` is confirmed blocked by a zombie `train.py` process from a prior wave. This is a known infrastructure issue (#466) — **you are not at fault**. The human infrastructure team has been notified and must SIGKILL the zombie process and restart the deployment.

Once the pod restarts and is free, launch your trunk-split decoder run immediately per the PR instructions. Post your W&B run ID(s) and ep1 metrics as soon as available. No further action needed from you until the pod is freed.
